[#21] [#8] As a user, I can login with an email and password

liamstevens111 opened the pull request:
- Close #21 
- Close #8 

## What happened 👀

- [x] Setup Base Adapter and Auth Adapter (includes client_id, client_secret) for performing requests with RequestManager
- [x] Setup environment variables for base URL and client and secret 
- [x] Perform login request and store `access_token` and `refresh_token` from response in LocalStorage 
- [x] Perform validation and handle API error and success responses on UI
- [x] Tests / Mocks for Login page, API requests + Adapters etc

Protecting survey pages and redirecting to homepage has been separated to another issue here: https://github.com/liamstevens111/react-survey-ic/issues/25

Maintaining a session by obtaining a new access token with refresh token been moved to separate ticket here: https://github.com/liamstevens111/react-survey-ic/issues/36

## Insight 📝

- Used global `localStorage` mock due to being client wide library and unable to be mocked naturally by Jest
- Used `nock` package for API request mocking for login page and AuthAdapter
- Used `setupTests` to set the values for `environment` variables in tests
- Real email/password used in cypress test, unsure on this one.

## Proof Of Work 📹

Tests pass

liamstevens111 committed:
[#25] Initial

liamstevens111 committed:
[#21] Base and Auth adapter for sending request, only includes login functionality wip

liamstevens111 committed:
[#21] Add UserToken helper module abstraction that uses LocalStorage for tokens

liamstevens111 committed:
[#21] Add test code for making request and storing token on client side - will remove

liamstevens111 committed:
[#21] Move base URL from auth adapter to base adapter as common for entirely of application

liamstevens111 committed:
[#21] Fix padding and add value prop to Input element. Make button type optional

liamstevens111 committed:
[#21] Allow form to send user/password to API for token. Use Form onSubmit event instead button click

liamstevens111 committed:
[#21] Add form error handling WIP

liamstevens111 committed:
Cleanup

liamstevens111 committed:
[#21] Change form error state from map of key/values to just a list due to uneeded complexity/issues

liamstevens111 committed:
[#21] Update package lock since pinning versions

liamstevens111 committed:
[#21] Use nock for mocking API response WIP

liamstevens111 committed:
[#21] Add tests for valid and invalid credientials

liamstevens111 committed:
[#21] Add env example file

liamstevens111 committed:
[#21] Add tests for LocalStorage and authAdapter

liamstevens111 committed:
[#21] Add cypress tests

liamstevens111 committed:
[#21] Commented out the two cypress tests failing on CI/CD for login with interceptors

hanam1ni commented on src/screens/Login/index.tsx:
```diff
@@ -0,0 +1,127 @@
+
+      navigate('/');
+    } catch (error) {
+      let errorMessage = 'There was a problem receiving a response from the server';
```
Should we add this base error message to the translation file?

liamstevens111 replied:
Yup, forgot that one :)

Added here https://github.com/liamstevens111/react-survey-ic/pull/35/commits/61d68f0692a140ef23d1976a0d9f0abb4d304771

hanam1ni commented on src/adapters/baseAdapter..ts:
```diff
@@ -0,0 +1,15 @@
+    return requestManager('GET', `${BaseAdapter.baseUrl}/${endpoint}`, params);
+  }
+
+  postRequest(endpoint: string, params: RequestParamsType) {
```
Should this be request option to be more accurate with the request manager?
```suggestion
  postRequest(endpoint: string, requestOption: RequestParamsType) {
```

liamstevens111 replied:
I used the naming conventions from `Compass` so I think should be default, but I can go with this as well

edit: changed here [9d0540f](https://github.com/liamstevens111/react-survey-ic/pull/35/commits/9d0540f069e38e276483c8c78f9c1822a03c36a1) 👍

hanam1ni reviewed (commented):
I get not found page when trying to access login path on [preview site](https://63f717ec7c8f5b7225e469ef--moonlit-buttercream-05cb2f.netlify.app/login). Not sure, if I am doing something wrong 🤔
<img width="722" alt="image" src="https://user-images.githubusercontent.com/14077479/221461176-e0ac2086-f4a5-4b9f-b300-2ce0c64300bc.png">

liamstevens111 commented:
> I get not found page when trying to access login path on [preview site](https://63f717ec7c8f5b7225e469ef--moonlit-buttercream-05cb2f.netlify.app/login). Not sure, if I am doing something wrong 🤔 <img alt="image" width="722" src="https://user-images.githubusercontent.com/14077479/221461176-e0ac2086-f4a5-4b9f-b300-2ce0c64300bc.png">

@hanam1ni Yes, I noticed this too so I waited until merged to main to see. I think this will resolve the issue:

https://stackoverflow.com/questions/50752350/page-not-found-when-trying-to-access-a-site-deployed-on-netlify

Works now, I made following changes: [9538bf5](https://github.com/liamstevens111/react-survey-ic/pull/35/commits/9538bf57e326bad1d08c828d7022f0c0e48e4cf2). It involved creating a _redirects file.

Login page: https://63fedb58e0e03327cdf00509--moonlit-buttercream-05cb2f.netlify.app/login

Edit: Woops, login page doesn't send request correctly as client id and secret not in environment, or the URL. I added to netlify but not sure I need in workflow too, dont think so.

liamstevens111 committed:
[#21] Change baseAdapter request function argument name from param to requestOption

liamstevens111 committed:
[#21] Add generic server error to translations

liamstevens111 committed:
[#21] Attempt to fix Netflix deploy on links not working

liamstevens111 committed:
[#21] Attempt to fix Netflix deploy on links not working

liamstevens111 committed:
[#21] Add artifact upload to test workflow for cypress failures

liamstevens111 committed:
[#21] Add artifact upload for failed cypress tests

rosle commented on src/screens/Login/index.tsx:
```diff
@@ -0,0 +1,127 @@
+            {t('login.forgot-password')}
+          </a>
+        </div>
+        <Button text={t('login.sign-in')} className="h-14 w-80" type={'submit'} disabled={formSubmitted} />
```
I think we could also pass like this?

```suggestion
        <Button text={t('login.sign-in')} className="h-14 w-80" type="submit" disabled={formSubmitted} />
```

rosle commented on src/screens/Login/index.tsx:
```diff
@@ -0,0 +1,127 @@
+        <Input
+          name="email"
+          label={t('login.email')}
+          type="text"
```
why don't we use `type="email"` 🤔 or we want to use custom validation here?

liamstevens111 replied:
yeah I didn't want the html5 default errors to appear on the page for the control (which I think is what would happen). What do you think? Is it not the preferred approach?

rosle commented on src/screens/Login/index.tsx:
```diff
@@ -0,0 +1,127 @@
+          className="my-3 block h-14 w-80"
+          onInputChange={handleEmailChange}
+        />
+        <div className="relative w-80">
```
Maybe we don't need `w-80` as we already set the width of the password input itself 💭

liamstevens111 replied:
This comment and previous rectified here 👍 

[9929fd1](https://github.com/liamstevens111/react-survey-ic/pull/35/commits/9929fd1fa4e5936c82c29767c84b6e3ff7898e12)

rosle commented on src/screens/Login/index.tsx:
```diff
@@ -0,0 +1,127 @@
+    <>
+      <img className="inline-block" src={logo} alt="logo" />
+      <p data-test-id="login-header" className="my-8 text-white opacity-50">
+        {t('login.sign-in')} to Nimble
```
Should we extract the whole text for translation? 🤔  (including to Nimble)

liamstevens111 replied:
I answered that in a previous PR - https://github.com/liamstevens111/react-survey-ic/pull/33#discussion_r1112694024

My thinking was that `Nimble` is the company name and I don't think we should "translate" the company name? I think it would be the same in all languages? Let me know if you think otherwise :)

liamstevens111 committed:
[#21] Remove redundant class and javascript curcly braces for props

liamstevens111 committed:
[#21] Fix artifact upload in workflow and add env variables to ensure cypress tests have API data

liamstevens111 commented:
I fixed cypress tests by adding environment variables to github workflow.

I added environment variables to Netlify but the `preview` is still showing "undefined" for URL during /login:

<img width="1497" alt="Screenshot 2023-03-08 at 11 24 44" src="https://user-images.githubusercontent.com/8955671/223619347-45058388-ff62-48d5-a3d5-1189d2024099.png">

Unsure at the moment, maybe I can address this in my other open PR https://github.com/liamstevens111/react-survey-ic/pull/34. Is that OK?

cc @hanam1ni
